internal/tempfile/tempfile_test.go

```diff
@@ -0,0 +1,180 @@
+package tempfile_test
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+
+	"github.com/buildkite/agent/v3/internal/tempfile"
+	"gotest.tools/v3/assert"
+)
+
+func TestNew(t *testing.T) {
+	t.Parallel()
+
+	f, err := tempfile.New()
+	assert.NilError(t, err, `New() = %v`, err)
+
+	t.Cleanup(func() {
+		assert.Check(t, f.Close() == nil, "failed to close file: %s", f.Name())
+		assert.Check(t, os.Remove(f.Name()) == nil, "failed to remove file: %s", f.Name())
+	})
+
+	assert.Assert(t, strings.HasPrefix(f.Name(), os.TempDir()))
+}
+
+func TestNewWithFilename(t *testing.T) {
+	t.Parallel()
+
+	f, err := tempfile.New(tempfile.WithName("foo.txt"))
+	assert.NilError(t, err, `New(WithName("foo.txt")) = %v`, err)
+
+	t.Cleanup(func() {
+		assert.Check(t, f.Close() == nil, "failed to close file: %s", f.Name())
+		assert.Check(t, os.Remove(f.Name()) == nil, "failed to remove file: %s", f.Name())
+	})
+
+	assert.Assert(t, strings.HasPrefix(f.Name(), os.TempDir()))
+}
+
+func TestNewWithDir(t *testing.T) {
+	t.Parallel()
+
+	f, err := tempfile.New(tempfile.WithDir("foo"))
+	assert.NilError(t, err, `New(WithDir("foo")) = %v`, err)
+
+	t.Cleanup(func() {
+		assert.Check(t, f.Close() == nil, "failed to close file: %s", f.Name())
+		assert.Check(t, os.Remove(f.Name()) == nil, "failed to remove file: %s", f.Name())
```

Comment:
Since `WithDir` creates `foo`, this cleanup ought to `os.RemoveAll(filepath.Join(os.TempDir(), "foo"))`

(Don't forget `TestNewClosedWithDir`)